test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -0,0 +1,148 @@
+     * on the 'Fail' button and the reason for the failure should be
+     * specified in the JDailog JTextArea.
+     *
+     * @param title           title of the Frame.
+     * @param instructions    specified instruction that user should follow.
+     * @param maxRowLength    number of visible rows of the JTextArea where the
+     *                        instruction is show.
+     * @param maxStringLength number of columns of the instructional JTextArea
+     * @param timeoutMinutes  timeout of the test where time is specified in
+     *                        minutes.
+     * @throws HeadlessException
+     * @throws InterruptedException
+     * @throws InvocationTargetException
+     */
+    public PassFailJFrame(String title, String instructions,
+                          int maxRowLength, int maxStringLength,
+                          int timeoutMinutes) throws HeadlessException,
+            InterruptedException, InvocationTargetException {
+        super(title);
+        this.maxRowLength = maxRowLength;
+        this.maxStringLength = maxStringLength;
+        this.timeoutMinutes = timeoutMinutes;
+
+        invokeAndWait(() -> {
+            setLayout(new BorderLayout());
+            instructionsText = new JTextArea(instructions, maxRowLength, maxStringLength);
+            instructionsText.setEditable(false);
+            instructionsText.setFocusable(false);
+            add(instructionsText, BorderLayout.NORTH);
+
+            JButton btnPass = new JButton("Pass");
+            btnPass.addActionListener((e) -> latch.countDown());
+
+            JButton btnFail = new JButton("Fail");
+            btnFail.addActionListener((e) -> getFailureReason());
+
+            JPanel buttonsPanel = new JPanel();
+            buttonsPanel.add(btnPass);
+            buttonsPanel.add(btnFail);
+
+            add(buttonsPanel, BorderLayout.SOUTH);
+            pack();
+            setLocation(10, 10);
+            setVisible(true);
+        });
+    }
+
+    /**
+     * Wait for the user decision i,e user selects pass or fail button.
+     * If user does not select pass or fail button then the test waits for
+     * the specified timeoutMinutes period and the test gets timeout.
+     *
+     * @throws InterruptedException
+     * @throws InvocationTargetException
+     */
+    public void awaitAndCheck() throws InterruptedException, InvocationTargetException {
+        boolean timeoutHappened = !latch.await(this.timeoutMinutes,
+                TimeUnit.MINUTES);
+        if (isEventDispatchThread()) {
+            dispose();
+        } else invokeAndWait(() -> dispose());
+
+        if (timeoutHappened) {
+            throw new RuntimeException("Test timed out!");
+        }
+        if (failed) {
+            throw new RuntimeException("Test failed! : " + testFailedReason);
+        }
+    }
+
+    public void getFailureReason() {
+        final JDialog dialog = new JDialog();
+        dialog.setTitle("Failure reason");
```

Comment:
I suggest making the dialog Toolkit-modal so that only the dialog can be interacted with.

```java
        final JDialog dialog = new JDialog(this, "Failure reason", TOOLKIT_MODAL);
```

At this time, I can click Fail again several times and I'll have several dialogs open. I can also click Pass which completes the test and disposes of the dialogs.

Reply:
Yes, I made the JDialog to be Modal dialog